documentation/proposals/Proposal - Multi-Backend Input.md

```diff
@@ -11,203 +11,683 @@ Proposal API for backend-agnostic, refactored Input via keyboards, mice, and con
+
+Note that the parties in this proposal are:
+- **developer**: The developer using Silk.NET.
+- **end user**: The user using an application using Silk.NET, and by extension the user inputting data into the application via Silk.NET.
+
+Cases where the **user** word is used without the **end** prefix can be assumed to be referring to the **developer** rather than the **end user**.
+
+# Usage Examples
+```cs
+IWindowHandlesSource someWindow = null!;
+var inputContext = someWindow.CreateInput();
+inputContext.Gamepads.ThumbstickMove += @event =>
+{
+    Console.WriteLine($"Thumbstick {@event.Index} moved from {@event.OldValue} to {@event.NewValue}");
+};
+var isButtonDown = inputContext.Gamepads.Any(gamepadState => gamepadState.Buttons[JoystickButton.A]);
+```
+```cs
+IWindowHandlesSource someWindow = null!;
+var inputContext = new InputContext();
+inputContext.Backends.Add(someWindow.CreateInputBackend());
+// in future:
+// inputContext.Backends.Add(new OpenXRInputBackend(...));
+```
+
+# Reference Implementation
+
+Similar to Windowing 3.0, a reference implementation will be included in the main `Silk.NET.Input` package which uses the same API or family of APIs as Windowing's reference implementation. This will be exposed by the `InputWindowExtensions` class.
+
+```cs
+public static class InputWindowExtensions
+{
+    public static IInputBackend CreateInputBackend(this WindowHandles window);
+    public static IInputBackend CreateInputBackend(this IWindowHandlesSource window);
+    public static InputContext CreateInput(this WindowHandles window);
+    public static InputContext CreateInput(this IWindowHandlesSource window);
+}
+```
+
+The `CreateInputBackend` will create an instance of the reference implementation for the given `WindowHandles`. The `IWindowHandlesSource` overloads just forward to the `WindowHandles` overload. This is because `ISurface` will implement `IWindowHandlesSource`, so the extension methods will be usable on an `ISurface` without having a hard reference between Windowing and Input.
+
+The `CreateInput` methods simply return an `InputContext` preconfigured with the backend created by `CreateInputBackend` for ease of use.
+
+Please see the Windowing 3.0 proposal for `IWindowHandlesSource` and `WindowHandles`.
+
+# Devices, Backends, and Contexts
+
+Input devices all inherit from a root interface.
+
+```cs
+public interface IInputDevice
+{
+    nint BackendId { get; }
+    string Name { get; }
+}
+```
+
+`BackendId` is an integral identifier for this device unique to this backend **only** - i.e. it is not globally unique, and if you have an input context with multiple backends there may be multiple devices with a particular `BackendId`. It may be a handle to memory, or it may be a generic integral index - this is implementation-defined.
```

Comment:
Can we also provide a global Id? in my book a GUID would make sense, as iirc we do not provide any guarantee the object reference stays the same either, so I feel providing a global ID (ie for storing what device the user selected or such?) would make sense.

Instead of a GUID the Input handle + device handle or something could also work (?)

Reply:
I suppose we could just make this globally unique, not a fan of GUIDs! Not specifying that this is a handle though, **INFORMATIVE TEXT:** because the way I'm picturing it now for devices that have an index instead of a handle will just do `backend handle + device index`

Reply:
```suggestion
`Id` is an globally-unique integral identifier for this device.
```